rclcpp/src/rclcpp/serialized_message.cpp

```diff
@@ -0,0 +1,134 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/serialized_message.hpp"
+
+#include <cstring>
+
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/logging.hpp"
+
+#include "rmw/types.h"
+
+namespace rclcpp
+{
+
+/// Object oriented version of rcl_serialized_message_t with destructor to avoid memory leaks
+SerializedMessage::SerializedMessage(const rcl_allocator_t & allocator)
+: SerializedMessage(0u, allocator)
+{}
+
+SerializedMessage::SerializedMessage(
+  size_t initial_capacity, const rcl_allocator_t & allocator)
+: rcl_serialized_message_t(rmw_get_zero_initialized_serialized_message())
+{
+  const auto ret = rmw_serialized_message_init(
+    this, initial_capacity, &allocator);
+  if (ret != RCL_RET_OK) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+SerializedMessage::SerializedMessage(const SerializedMessage & serialized_message)
+: SerializedMessage(static_cast<const rcl_serialized_message_t &>(serialized_message))
+{}
+
+SerializedMessage::SerializedMessage(const rcl_serialized_message_t & serialized_message)
+: rcl_serialized_message_t(rmw_get_zero_initialized_serialized_message())
+{
+  const auto ret = rmw_serialized_message_init(
+    this, serialized_message.buffer_capacity, &serialized_message.allocator);
+  if (ret != RCL_RET_OK) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  // do not call memcpy if the pointer is "static"
+  if (buffer != serialized_message.buffer) {
+    std::memcpy(buffer, serialized_message.buffer, serialized_message.buffer_length);
+  }
+  buffer_length = serialized_message.buffer_length;
+}
+
+SerializedMessage::SerializedMessage(SerializedMessage && serialized_message)
+: SerializedMessage(
+    std::forward<rcl_serialized_message_t>(
+      static_cast<rcl_serialized_message_t &&>(serialized_message)))
+{}
+
+SerializedMessage::SerializedMessage(rcl_serialized_message_t && serialized_message)
+: rcl_serialized_message_t(serialized_message)
+{
+  // reset buffer to prevent double free
+  serialized_message = rmw_get_zero_initialized_serialized_message();
+}
+
+SerializedMessage & SerializedMessage::operator=(const SerializedMessage & other)
+{
+  *this = static_cast<const rcl_serialized_message_t &>(other);
+
+  return *this;
+}
+
+SerializedMessage & SerializedMessage::operator=(const rcl_serialized_message_t & other)
+{
+  *this = static_cast<SerializedMessage>(rmw_get_zero_initialized_serialized_message());
+
+  const auto ret = rmw_serialized_message_init(
+    this, other.buffer_capacity, &other.allocator);
+  if (ret != RCL_RET_OK) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  // do not call memcpy if the pointer is "static"
+  if (buffer != other.buffer) {
+    std::memcpy(buffer, other.buffer, other.buffer_length);
+  }
+  buffer_length = other.buffer_length;
```

Comment:
I think that's done deliberately.

Reply:
Sure, but it doesn't sound like a `class`, but more like an `struct` according to our guidelines:
https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes.

I would prefer to continue using a class, use composition (with the rcl object as a private member), and add the needed getters/setters.
The other option is to use a `struct` instead of a `class`, as there aren't currently methods in the definition (more than initialization/constructors/destructors).

Reply:
I don't feel super strongly about whether we use inheritance or composition here.
I do believe though it should be a class. I could see that we enhance this class further with methods like `resize()` or `reserve()` to make it more like a container.

Reply:
Sounds good.

I have a strong preference about only having `private` members on classes, but I won't block on the change.

Reply:
I'm just browsing this after the fact while waiting on a build, but I agree that inheriting from a C struct is not a great idea. Having a private/protected member that is the rcl type would be safer and more future proof (changing this in the future becomes harder because we don't have overloading or the opportunity to deprecate things due to member access).